File: _extensions/code-mark/code-mark.lua
-- Highlights **part of a code line** — a red box around a piece of text.
--
--     ```{.go mark="Rows - 1"}
--     ```{.go mark="5:Rows - 1"}                  -- search line 5 only
--     ```{.go mark="1:(col int);5:Rows - 1"}      -- several, separated by ;
--
-- Inside `.explain-code` the mark belongs on the explanation step rather than on
-- the block; `explain-code.lua` collects those and leaves a `mark-steps` here —
-- one mark per step, separated by `|`, in the same segmentation as
-- `code-line-numbers`:
--
--     :::: {lines="5" mark="Rows - 1"}
--     We count from the bottom up.
--     ::::
--
-- Without this there is no way to point at *this* spot: `code-line-numbers` and
-- the `lines="…"` steps of `.explain-code` can only take whole lines, and so can
-- Quarto's `code-annotations` (the issue asking for less, #11154, is closed as
-- "not planned").
--
-- Why the box is drawn at runtime instead of here: pandoc produces the syntax
-- highlighting itself, a filter only ever sees the raw code text. The token
-- spans a mark has to run across (`fmt` `.` `Errorf` are three spans) come into
-- existence after us. In the DOM, wrapping them is a standard problem — that is
-- what code-mark.js does.
--
-- So this filter does three things:
--   1. It checks **at build time** that the text is there at all. A typo in
--      `mark=` would otherwise stay silent.
--   2. It attaches the JS and CSS, but only when something is actually marked.
--   3. For LaTeX it removes the attributes again; there is no `data-*` there.

-- Quarto hands pandoc a temporary intermediate file, so `PANDOC_STATE.input_files`
-- points into /tmp. `quarto.doc.input_file` knows the real path but is not there
-- in every version — hence the fallback.
local function input_file()
	local ok, path = pcall(function()
		return quarto.doc.input_file
	end)
	if ok and type(path) == "string" and path ~= "" then
		return path
	end
	if PANDOC_STATE ~= nil and PANDOC_STATE.input_files ~= nil and #PANDOC_STATE.input_files > 0 then
		return PANDOC_STATE.input_files[1]
	end
	return "<unknown file>"
end

-- Aborts the render with a readable message.
local function fail(el, reason, hint)
	local first_line = (el.text:match("^%s*([^\n]+)") or ""):match("^%s*(.-)%s*$")
	local spec = el.attributes["mark"] or el.attributes["mark-steps"] or ""
	io.stderr:write("\n")
	io.stderr:write("=== code-mark: " .. reason .. " ===\n")
	io.stderr:write("  File       : " .. input_file() .. "\n")
	io.stderr:write("  Code block : " .. first_line .. "\n")
	io.stderr:write("  mark       : " .. spec .. "\n")
	io.stderr:write("\n  " .. hint .. "\n\n")
	os.exit(1)
end

-- Splits at `sep`, but not at an occurrence escaped with `\`.
--
-- Needed because both separators show up in real code: `;` in the head of a
-- classic `for` loop, `|` in every `||`. A `\` before anything else stays a `\` —
-- each level unescapes only its own separator, so the two (first `|`, then `;`)
-- do not get in each other's way.
--
-- Of the two, only `;` is visible to authors, and in a .qmd it needs **two**
-- backslashes (`\\;`): pandoc's attribute parser eats one. The `|` is only ever
-- split in the internally generated `mark-steps`, which never passes that parser
-- — explain-code.lua escapes it there itself.
local function split_escaped(s, sep)
	local out = {}
	local buf = {}
	local i = 1
	while i <= #s do
		local c = s:sub(i, i)
		if c == "\\" and s:sub(i + 1, i + 1) == sep then
			table.insert(buf, sep)
			i = i + 2
		elseif c == sep then
			table.insert(out, table.concat(buf))
			buf = {}
			i = i + 1
		else
			table.insert(buf, c)
			i = i + 1
		end
	end
	table.insert(out, table.concat(buf))
	return out
end

-- "5:Rows - 1" -> { line = 5, needle = "Rows - 1" }
-- "Rows - 1"   -> { line = nil, needle = "Rows - 1" }
local function parse_entry(part)
	local line, rest = part:match("^%s*(%d+)%s*:(.*)$")
	if line then
		return { line = tonumber(line), needle = rest }
	end
	return { line = nil, needle = (part:gsub("^%s+", "")) }
end

local function split_lines(text)
	local out = {}
	for line in (text .. "\n"):gmatch("([^\n]*)\n") do
		table.insert(out, line)
	end
	-- the fence contributes an empty trailing line
	if out[#out] == "" then
		table.remove(out)
	end
	return out
end

-- Checks one `mark` spec against the displayed code.
local function validate(cb, spec, lines)
	for _, part in ipairs(split_escaped(spec, ";")) do
		if part:match("%S") then
			local e = parse_entry(part)
			if e.needle == "" then
				fail(cb, "empty search text", "Expected `mark=\"text\"` or `mark=\"line:text\"`.")
			end

			local haystack
			if e.line then
				if lines[e.line] == nil then
					fail(cb, "no line " .. e.line .. " (the block has " .. #lines .. ")",
						"Line numbers count the **displayed** code, like `lines=` in `.explain-code`.")
				end
				haystack = lines[e.line]
			else
				haystack = cb.text
			end

			if not haystack:find(e.needle, 1, true) then
				fail(cb, "text not found: " .. e.needle,
					"The search is literal, not a pattern. A typo, or did the code change?\n" ..
					"  A `;` that belongs to the search text is written `\\\\;` in the .qmd — two backslashes,\n" ..
					"  pandoc eats one while reading the attribute. A `|` needs nothing.\n" ..
					"  Without this check the slide would simply come out unmarked, and you would\n" ..
					"  find out while presenting.")
			end
		end
	end
end

function CodeBlock(cb)
	local spec = cb.attributes["mark"]
	local steps = cb.attributes["mark-steps"]
	if (spec == nil or spec == "") and (steps == nil or steps == "") then
		return nil
	end

	if quarto.doc.is_format("latex") then
		cb.attributes["mark"] = nil
		cb.attributes["mark-steps"] = nil
		return cb
	end

	local lines = split_lines(cb.text)

	if spec and spec ~= "" then
		validate(cb, spec, lines)
	end
	if steps and steps ~= "" then
		for _, step in ipairs(split_escaped(steps, "|")) do
			validate(cb, step, lines)
		end
	end

	quarto.doc.add_html_dependency({
		name = "code-mark",
		version = "0.1.0",
		scripts = { "code-mark.js" },
		stylesheets = { "code-mark.css" },
	})

	-- Quarto turns the unknown attributes into `data-mark` and `data-mark-steps`
	-- on the outer `div.sourceCode` by itself — which is exactly where
	-- code-mark.js looks for them.
	return cb
end

File: _extensions/code-mark/code-mark.js
// Draws the red box around the pieces of text named in `mark="…"`.
//
// The filter (code-mark.lua) has already checked the search text at build time
// and left the attribute as `data-mark` on the outer `div.sourceCode`. What is
// left is the wrapping, in the DOM:
//
//   <span id="cb1-5"> … Rows <span class="op">-</span> <span class="dv">1</span> … </span>
//
// A hit therefore regularly runs across several token spans and cannot be put
// into *one* element. It is split into one wrapper per text node instead; the
// first gets `code-mark-start`, the last `code-mark-end`. Together with the
// edges from code-mark.css that reads as one continuous box.
//
// On slides this runs **after** Quarto's line-highlight plugin: that clones the
// whole <code> node once per `|` step, and the clones sit in step order in the
// DOM. `data-mark-steps` hangs off exactly that: one mark per step, in the same
// segmentation as `code-line-numbers`. From `.explain-code` it comes for free —
// there `mark=` sits on the explanation step.

(function () {
	// Splits at `sep`, but not at an occurrence escaped with `\`.
	// Both separators show up in real code (`;` in a for head, `|` in `||`);
	// to mark one, write `\;` or `\|`. See code-mark.lua.
	function splitEscaped(s, sep) {
		var out = [];
		var buf = "";
		for (var i = 0; i < s.length; i++) {
			var c = s.charAt(i);
			if (c === "\\" && s.charAt(i + 1) === sep) {
				buf += sep;
				i++;
			} else if (c === sep) {
				out.push(buf);
				buf = "";
			} else {
				buf += c;
			}
		}
		out.push(buf);
		return out;
	}

	// "5:Rows - 1" -> {line: 5, needle: "Rows - 1"} · "Rows - 1" -> {line: null, …}
	function parseSpec(spec) {
		return splitEscaped(spec, ";")
			.map(function (part) {
				var m = /^\s*(\d+)\s*:(.*)$/.exec(part);
				if (m) return { line: parseInt(m[1], 10), needle: m[2] };
				return { line: null, needle: part.replace(/^\s+/, "") };
			})
			.filter(function (e) {
				return e.needle !== "";
			});
	}

	function textNodes(root) {
		var out = [];
		var walker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT, null);
		var n;
		while ((n = walker.nextNode())) {
			// The line numbers hang off an empty <a> as ::before — its text nodes
			// (if any) are not part of the code.
			if (!(n.parentNode && n.parentNode.tagName === "A")) out.push(n);
		}
		return out;
	}

	// Wraps every occurrence of `needle` below `root`.
	function wrapOccurrences(root, needle) {
		var nodes = textNodes(root);
		if (!nodes.length) return;

		var text = nodes
			.map(function (n) {
				return n.nodeValue;
			})
			.join("");

		var hits = [];
		var from = 0;
		var at;
		while ((at = text.indexOf(needle, from)) !== -1) {
			hits.push([at, at + needle.length]);
			from = at + needle.length;
		}
		if (!hits.length) return;

		// Start offset of each node within the joined text.
		var starts = [];
		var pos = 0;
		nodes.forEach(function (n) {
			starts.push(pos);
			pos += n.nodeValue.length;
		});

		// Back to front, so the splits do not shift the offsets of the hits that
		// are still pending.
		hits.reverse().forEach(function (hit) {
			var hs = hit[0];
			var he = hit[1];
			for (var i = nodes.length - 1; i >= 0; i--) {
				var ns = starts[i];
				var ne = ns + nodes[i].nodeValue.length;
				if (ne <= hs || ns >= he) continue; // node outside the hit

				var node = nodes[i];
				var localStart = Math.max(hs - ns, 0);
				var localEnd = Math.min(he - ns, node.nodeValue.length);

				var mid = node;
				if (localEnd < mid.nodeValue.length) mid.splitText(localEnd);
				if (localStart > 0) mid = mid.splitText(localStart);

				var span = document.createElement("span");
				span.className = "code-mark";
				if (ns + localStart === hs) span.classList.add("code-mark-start");
				if (ns + localEnd === he) span.classList.add("code-mark-end");
				mid.parentNode.insertBefore(span, mid);
				span.appendChild(mid);
			}
		});
	}

	function lineSpans(code) {
		return Array.prototype.filter.call(code.children, function (el) {
			return el.tagName === "SPAN" && /^cb\d+-\d+/.test(el.id || "");
		});
	}

	function applyTo(code, entries) {
		if (code.dataset.codeMarkDone) return;
		code.dataset.codeMarkDone = "1";
		var lines = lineSpans(code);
		entries.forEach(function (e) {
			if (e.line) {
				if (lines[e.line - 1]) wrapOccurrences(lines[e.line - 1], e.needle);
			} else {
				wrapOccurrences(code, e.needle);
			}
		});
	}

	function run() {
		// Collect first, then apply once per <code>: a block can carry both
		// `data-mark` (valid for every step) and `data-mark-steps`, and wrapping
		// twice would nest the hits.
		var byCode = new Map();
		function add(code, entries) {
			if (!code || !entries.length) return;
			byCode.set(code, (byCode.get(code) || []).concat(entries));
		}

		// On the whole block: valid for every step.
		document.querySelectorAll("div.sourceCode[data-mark]").forEach(function (div) {
			var entries = parseSpec(div.getAttribute("data-mark"));
			div.querySelectorAll("pre code").forEach(function (code) {
				add(code, entries);
			});
		});

		// Per step: the i-th <code> is the i-th step. If that step is missing from
		// the DOM (because the line-highlight plugin did not clone at all, say),
		// the mark is skipped silently — it would have no home.
		document.querySelectorAll("div.sourceCode[data-mark-steps]").forEach(function (div) {
			var steps = splitEscaped(div.getAttribute("data-mark-steps"), "|");
			var codes = div.querySelectorAll("pre code");
			steps.forEach(function (step, i) {
				add(codes[i], parseSpec(step));
			});
		});

		byCode.forEach(function (entries, code) {
			applyTo(code, entries);
		});
	}

	function start() {
		if (window.Reveal && typeof window.Reveal.on === "function") {
			if (window.Reveal.isReady && window.Reveal.isReady()) run();
			else window.Reveal.on("ready", run);
		} else {
			run();
		}
	}

	if (document.readyState === "loading") {
		document.addEventListener("DOMContentLoaded", start);
	} else {
		start();
	}
})();
